sleuth-camel-core/src/main/java/com.playtika.sleuth.camel/CreatedEventNotifier.java

```diff
@@ -0,0 +1,101 @@
+ * SOFTWARE.
+ */
+
+package com.playtika.sleuth.camel;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.camel.Endpoint;
+import org.apache.camel.Exchange;
+import org.apache.camel.management.event.ExchangeCreatedEvent;
+import org.apache.camel.support.EventNotifierSupport;
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.SpanTextMap;
+import org.springframework.cloud.sleuth.Tracer;
+import org.springframework.cloud.sleuth.instrument.messaging.MessagingSpanTextMapExtractor;
+import org.springframework.cloud.sleuth.instrument.messaging.MessagingSpanTextMapInjector;
+import org.springframework.cloud.sleuth.util.SpanNameUtil;
+
+import java.util.EventObject;
+
+import static com.playtika.sleuth.camel.SleuthCamelConstants.FROM_CAMEL;
+
+@Slf4j
+@AllArgsConstructor
+public class CreatedEventNotifier extends EventNotifierSupport {
+
+    private static final String MESSAGE_COMPONENT = "camel";
+
+    private final Tracer tracer;
+    private final MessagingSpanTextMapExtractor spanExtractor;
+    private final MessagingSpanTextMapInjector spanInjector;
+
+    @Override
+    public void notify(EventObject event) {
+        log.trace("Caught ExchangeCreatedEvent, processing...");
+        ExchangeCreatedEvent exchangeCreatedEvent = (ExchangeCreatedEvent) event;
+        Exchange exchange = exchangeCreatedEvent.getExchange();
+        Endpoint endpoint = exchange.getFromEndpoint();
+        CamelMessageTextMap carrier = new CamelMessageTextMap(exchange.getIn());
+
+        Span span = getSpan(carrier, endpoint);
+        span.tag(FROM_CAMEL, "true");
+
+        spanInjector.inject(span, carrier);
+    }
+
+    private Span getSpan(CamelMessageTextMap carrier, Endpoint endpoint) {
+        Span spanFromMessage = buildSpan(carrier);
+        if (spanFromMessage != null) {
+            log.debug("Built span from message - {}. Assuming it is remote one. Continuing it...", spanFromMessage);
+            tracer.continueSpan(spanFromMessage);
+            spanFromMessage.logEvent(Span.SERVER_RECV);
+            return spanFromMessage;
+        } else {
+            String name = getSpanName(endpoint);
+            Span newSpan = tracer.createSpan(name);
+            newSpan.logEvent(Span.CLIENT_SEND);
+            log.debug("Message doesn't contain span data, new one was created - {}.", newSpan);
+            return newSpan;
+        }
+    }
+
+    private String getSpanName(Endpoint endpoint) {
+        return SpanNameUtil.shorten(MESSAGE_COMPONENT + "::" + endpoint.getEndpointKey());
+    }
+
+    private Span buildSpan(SpanTextMap carrier) {
+        try {
+            return this.spanExtractor.joinTrace(carrier);
+        } catch (Exception e) {
+            log.error("Exception occurred while trying to extract span from carrier.", e);
+            return null;
+        }
+    }
+
+    @Override
+    public boolean isEnabled(EventObject event) {
```

Comment:
isExchangeCreated()

Reply:
it's 3rd party method from camel `EventNotifier`